hal/network/ncp_client/quectel/quectel_ncp_client.cpp

```diff
@@ -1346,22 +1373,24 @@ int QuectelNcpClient::enterDataMode() {
     // 4) Switch to command mode; otherwise return to step 1)
     // NOTE: we are ignoring step 1 and are acting more optimistic on the first attempt
     // Subsequent attempts to exit data mode follow this step.
-    const int attempts = 5;
     bool responsive = false;
-    for (int i = 0; i < attempts; i++) {
+
+    // Initialize AT parser
+    auto parserConf = AtParserConfig()
+            .stream(muxerDataStream_.get())
+            .commandTerminator(AtCommandTerminator::CRLF);
+    dataParser_.destroy();
+    CHECK(dataParser_.init(std::move(parserConf)));
+
+    for (int i = 0; i < DATA_MODE_BREAK_ATTEMPTS; i++) {
+        // Also try toggling DTR as +++ may sometimes fail.
+        exitDataModeWithDtr();
         // Send data mode break
         const char breakCmd[] = "+++";
         muxerDataStream_->write(breakCmd, sizeof(breakCmd) - 1);
         skipAll(muxerDataStream_.get(), 1000);
 
-        // Initialize AT parser
-        auto parserConf = AtParserConfig()
-                .stream(muxerDataStream_.get())
-                .commandTerminator(AtCommandTerminator::CRLF);
-        dataParser_.destroy();
-        CHECK(dataParser_.init(std::move(parserConf)));
```

Comment:
I'd reset the parser here.

Reply:
It should be in a cleanly initialized state here, right? Or more like now in https://github.com/particle-iot/device-os/pull/2300/files/c1cd1ed590d079711d469117684f0e0577195bc0#diff-186ce5679947e07e5ce6c3d4fce9cdce5ebd61e4c315bcf0658db6106ea54299R1383

Reply:
Yeah, but we then reuse the same parser instance in a loop and ignore errors until the maximum number of attempts is reached. I don't think calling `dataParser_.reset()` would necessarily make a difference here, I just noticed that we usually call it in similar cases in other parts of the code.

Reply:
Added `dataParser_.reset()`